### DISCUSSION TITLE
How do I use the keyword extraction feature?

### DISCUSSION THREAD

**Discussion by ethindp** [tree-sitter/tree-sitter#2578] [Q&A]
So I'm trying to use the keyword extraction feature, but tree-sitter is claiming that I can't use a non-terminal symbol. (The rule is solely a regex.) I used the grex tool to create a regular expression for all the keywords in the language:

```js
    word: $ => $.keywords,

    rules: {
        keywords: $ => /(?:(?:synchroniz|alias|tagg)ed|o(?:verriding|[fr])|(?:(?:excep|func)tio|retur|begi|the)n|(?:interfac|(?:termin|separ)at|requeu|r(?:ever|ai)s|subtyp|declar|whil|cas|som|typ|us)e|p(?:r(?:otected|agma)|r(?:ocedur|ivat)e|a(?:rallel|ckage))|(?:constan|a(?:ccep|bor)|exi|ou|no)t|abstract|re(?:names|m)|g(?:eneric|oto)|limited|(?:acces|digit|other)s|select|(?:recor|an|mo)d|e(?:ls(?:if|e)|nd)|(?:(?:arr|del)a|entr)y|delta|range|(?:unti|nul)l|body|loop|task|w(?:hen|ith)|abs|a(?:ll|t)|new|for|xor|in|do|i[fs])/,
```

I've tried just a `choice(...)` over all keywords and it (still) doesn't like it. And there doesn't seem to be a negative/positive lookahead function for me to say "identifiers can be this but they shouldn't match these words". How do I use this feature properly?

**Comment by amaanq** [comment]
lookahead/behind/around is *not* supported. 

The word rule is not for keywords, but for identifiers or the like

**Comment by ethindp** [comment]
Okay.... I've set it to the identifier rule and it still says it can't be used as a word token (even though I rewrote the entire rule to solely use a regex).

**Comment by amaanq** [comment]
post your grammar

**Comment by ethindp** [comment]
Uh..... Alright, but it's quite a long file, so I've uploaded it.
[grammar.txt](https://github.com/tree-sitter/tree-sitter/files/12446826/grammar.txt)

**Comment by amaanq** [comment]
Identifier as your first rule makes that the top-level rule...

There seems to be a few misunderstandings of how to write a tree-sitter grammar there, I would look at upstream grammars for  examples of good grammars, and note that your first rule must be the "program" or "source_file" (or whatever) rule that signifies how to parse an entire file.

**Comment by ethindp** [comment]
So order of the rules does matter? Or is there a way I can explicitly specify the top-level one? (I'd prefer not to reorder all the rules, that'd be a PITA given the shear number of them.) Could I just move the top-level rule to the first rule?

**Reply by amaanq** [reply]
yeah, only the first rule matters, and ordering of terminal rules determines lexical precedence (don't worry about this if you have no idea what that means)

just a good rule of thumb, place all terminals like identifiers at the bottom.

**Reply by sogaiu** [reply]
If I'm following the conversation appropriately, I was also [bitten by the order thing](https://github.com/sogaiu/tree-sitter-clojure/blob/6e41628e9d18b19caea1cb1d72aae4ccff5bdfe9/grammar.js#L249).  I haven't seen this documented and if that's the case, perhaps it would be better if it were.

**Reply by ahlinc** [reply]
https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens
https://tree-sitter.github.io/tree-sitter/creating-parsers#using-precedence

**Reply by sogaiu** [reply]
Perhaps I've misunderstood the context of the discussion.

I was referring to the order of the first rule mattering -- that you have to put the start rule as the first thing in `rules`.

I looked at the linked sections but I did not find mention of:

> only the first rule matters

The closest thing I see is:

> Rule Order - If none of the above criteria can be used to select one token over another, Tree-sitter will prefer the token that appears earlier in the grammar.

But is this not about token precedence?  It doesn't seem to me be to be generally about rule order.  Or does it?

Sorry if what I'm referring to is not along the lines of this discussion or I've misunderstood somehow (^^;

**Reply by ahlinc** [reply]
@sogaiu you're right, there is no direct mention about importance of the first rule, it only exists in TS own tests.

To don't reorder rules and specify an entry point it's possible to do:
```js
grammar({
  inline: $ => [
    $._real_entry_point_rule,
  ],
  rules: $ => {
    entry_point_rule:  $ => $._real_entry_point_rule,

    // ...

   _real_entry_point_rule: $ => // ...
})
```
